crates/matrix-sdk-base/src/read_receipts.rs

```diff
@@ -0,0 +1,395 @@
+use matrix_sdk_common::deserialized_responses::SyncTimelineEvent;
+use ruma::{
+    events::{
+        poll::{start::PollStartEventContent, unstable_start::UnstablePollStartEventContent},
+        receipt::{ReceiptThread, ReceiptType},
+        room::message::Relation,
+        AnySyncMessageLikeEvent, AnySyncTimelineEvent, OriginalSyncMessageLikeEvent,
+        SyncMessageLikeEvent,
+    },
+    serde::Raw,
+    EventId, OwnedEventId, RoomId, UserId,
+};
+use tracing::{field::display, instrument, trace};
+
+use super::BaseClient;
+use crate::{error::Result, store::StateChanges, RoomInfo};
+
+/// Provider for timeline events prior to the current sync.
+pub trait PreviousEventsProvider: Send + Sync {
+    /// Returns the list of known timeline events, in sync order, for the given
+    /// room.
+    // TODO: return a reference or some kind of iterator
+    fn for_room(&self, room_id: &RoomId) -> Vec<SyncTimelineEvent>;
+}
+
+impl PreviousEventsProvider for () {
+    fn for_room(&self, _: &RoomId) -> Vec<SyncTimelineEvent> {
+        Vec::new()
+    }
+}
+
+#[instrument(skip_all, fields(room_id))]
+pub(crate) async fn compute_notifications(
+    client: &BaseClient,
+    changes: &StateChanges,
+    previous_events_provider: &dyn PreviousEventsProvider,
+    new_events: &[SyncTimelineEvent],
+    room_info: &mut RoomInfo,
+) -> Result<()> {
+    // Only apply the algorithm to encrypted rooms, since unencrypted rooms' unread
+    // notification counts ought to be properly computed by the server.
+    if !room_info.is_encrypted() {
+        return Ok(());
+    }
+
+    tracing::Span::current().record("room_id", display(&room_info.room_id));
+
+    let user_id = &client.session_meta().unwrap().user_id;
+    let prev_latest_receipt_event_id = room_info.latest_read_receipt_event_id.clone();
+
+    if let Some(receipt_event) = changes.receipts.get(room_info.room_id()) {
+        trace!("Got a new receipt event!");
+
+        // Find a private or public read receipt for the current user.
+        let mut receipt_event_id = None;
+        if let Some((event_id, receipt)) =
+            receipt_event.user_receipt(user_id, ReceiptType::ReadPrivate)
+        {
+            if receipt.thread == ReceiptThread::Unthreaded || receipt.thread == ReceiptThread::Main
+            {
+                receipt_event_id = Some(event_id.to_owned());
+            }
+        } else if let Some((event_id, receipt)) =
+            receipt_event.user_receipt(user_id, ReceiptType::Read)
+        {
+            if receipt.thread == ReceiptThread::Unthreaded || receipt.thread == ReceiptThread::Main
+            {
+                receipt_event_id = Some(event_id.to_owned());
+            }
+        }
```

Comment:
I wonder if we couldn't use the result of `receipt_event.user_receipt(user_id, ReceiptType::ReadPrivate).or_else(|| receipt_event.user_receipt(user_id, ReceiptType::Read))` to simplify the 2 blocks into a single one.

Proposed modification (not tested, maybe the type system or the borrow checker will invalidate it):

```rust
if let Some((event_id, receipt)) = receipt_event.user_receipt(user_id, ReceiptType::ReadPrivate).or_else(|| receipt_event.user_receipt(user_id, ReceiptType::Read)) {
    if receipt.thread == ReceiptThread::Unthreaded || receipt.thread == ReceiptThread::Main {
        receipt_event_id = Some(event_id.to_owned());
    }
}
```

Reply:
nice!